### PR TITLE
[feat,fix] Logger overhaul, shift to Python logger

### PR DESCRIPTION
**Pull request by apsdehal**
- All logging uses python based logging after this
- Colored logging support inspired from d2
- Wherever touched moved to format strings
- Adds options for colored logging
- Moves torch version and cuda devices to top
- Moves all print and writer statements to logger
- Also, fixes the long failing windows tests for logger

What are the benefits of doing this?
- All libraries which use python logging will receives our logs now and
we receive theirs
- Proper division between stderr and stdout

Test Plan:

Tested locally

Reviewer suggestions:
This PR needs to be reviewed somewhat carefully to make sure I don't miss anything and as it touches all of the MMF.

### CODE REVIEW

File: mmf/configs/defaults.yaml [hunk 1]
```diff
@@ -33,8 +33,8 @@ training:
     # Whether MMF should log or not, Default: False, which means
     # mmf will log by default
     should_not_log: false
-    # Whether to capture stdout logs by the logger. Default: True
-    stdout_capture: true
```
Comment (vedanuj) [inline mmf/configs/defaults.yaml:37]: Are we not capturing stdout logs anymore or not having an option to do so? Or did I miss something?

File: mmf/utils/build.py [hunk 2]
```diff
@@ -63,7 +63,7 @@ def build_model(config):
     model_class = registry.get_model_class(model_name)
 
     if model_class is None:
-        registry.get("writer").write("No model registered for name: %s" % model_name)
+        raise RuntimeError(f"No model registered for name: {model_name}")
```
Comment (vedanuj) [inline mmf/utils/build.py:66]: Nice!! 👍

File: mmf/utils/logger.py [hunk 3]
```diff
@@ -1,174 +1,251 @@
...
+    distributed_rank = get_rank()
+    handlers = []
+
+    if distributed_rank == 0:
+        logger.setLevel(logging.INFO)
+        ch = logging.StreamHandler(stream=sys.stdout)
+        ch.setLevel(logging.INFO)
+        if color:
+            formatter = ColorfulFormatter(
+                colored("%(asctime)s | %(name)s: ", "green") + "%(message)s",
```
Comment (vedanuj) [inline mmf/utils/logger.py:106]: Missing `%(levelname)s`

File: mmf_cli/run.py [hunk 4]
```diff
@@ -31,12 +35,17 @@ def main(configuration, init_distributed=False, predict=False):
...
     config = build_config(configuration)
 
-    # Logger should be registered after config is registered
-    registry.register("writer", Logger(config, name="mmf.train"))
+    setup_logger(
+        color=config.training.colored_logs, disable=config.training.should_not_log
+    )
+    logger = logging.getLogger("mmf_cli.run")
+    logger.info(f"Torch version: {torch.__version__}")
+    log_device_names()
```
Comment (vedanuj) [inline mmf_cli/run.py:46]: Nice 👍

File: tests/utils/test_logger.py [hunk 5]
```diff
@@ -28,23 +29,34 @@ def setUpClass(cls) -> None:
...
 
     @classmethod
     def tearDownClass(cls) -> None:
         # Cleanup temp working dir.
-        for handler in cls.writer._logger.handlers:
+        for handler in cls.writer.handlers:
             handler.close()
         if cls._tmpdir is not None:
             # set ignore_errors as Windows throw error due to Permission
             shutil.rmtree(cls._tmpdir, ignore_errors=True)
```
Comment (vedanuj) [inline tests/utils/test_logger.py:43]: Not sure what you mean by `long failing windows logger tests` in your PR description. Those were already passing earlier. One issue was regarding permissions removing the logger folder. If that is fixed by these changes, then we can remove `ignore_errors=True` here.

File: mmf/configs/defaults.yaml (hunk 1 above)
Comment (apsdehal) [reply mmf/configs/defaults.yaml:37]: We are capturing stdout always. There is a should_not_log option that basically does the same thing as setting this to false.

File: mmf/utils/logger.py (hunk 3 above)
Comment (apsdehal) [reply mmf/utils/logger.py:106]: Colorful Formatter has color for log levels. INFO is green, and when warning or error are logged they are appended on front in red color. Check the source for Colorful Formatter.

File: tests/utils/test_logger.py (hunk 5 above)
Comment (apsdehal) [reply tests/utils/test_logger.py:43]: The log files with ":" were some how still failing with OSError. I fixed them by changing ":" to "_".

File: mmf/configs/defaults.yaml (hunk 1 above)
Comment (vedanuj) [reply mmf/configs/defaults.yaml:37]: `stdout_capture` here means capturing stdout to log files. It is different from `should_not_log`.

File: mmf/utils/logger.py [hunk 6]
```diff
@@ -1,174 +1,251 @@
 # Copyright (c) Facebook, Inc. and its affiliates.
 
 import collections
```
Comment (vedanuj) [inline mmf/utils/logger.py:3]: Will be good to add a comment here that most of this is adapted from Detectron2.

File: tests/utils/test_logger.py (hunk 5 above)
Comment (vedanuj) [reply tests/utils/test_logger.py:43]: Thanks. PR #352 did fix some log file names. Maybe it missed some.

File: mmf/configs/defaults.yaml (hunk 1 above)
Comment (apsdehal) [reply mmf/configs/defaults.yaml:37]: Yes, discussed offline. The option isn't useful anymore.

**Pull request closed without merging**